ansible_base/jwt_consumer/common/auth.py

```diff
@@ -223,9 +240,19 @@ def process_rbac_permissions(self):
                 object_data = self.token['objects'][object_type][index]
                 resource, obj = self.get_or_create_resource(object_type, object_data)
                 if resource is not None:
-                    rd.give_permission(self.user, obj)
+                    assignment = rd.give_permission(self.user, obj)
+                    role_diff = role_diff.exclude(pk=assignment.pk)
                     logger.info(f"Granted user {self.user.username} role {object_role_name} to object {obj.name} with ansible_id {object_data['ansible_id']}")
 
+        # Remove all permissions not authorized by the JWT
+        for role_assignment in role_diff:
+            rd = role_assignment.role_definition
+            object_role = role_assignment.object_role
+            if object_role:
+                rd.remove_permission(self.user, object_role.content_object)
```

Comment:
Performance-wise, you can also reference `role_assignment.content_object`. It's the same content_object. This would avoid an additional query.